### PR TITLE
Send ping every second truly async to all nodes who does not have any…

### PR DESCRIPTION
**Pull request by baldersheim**
… pending pings.

@havardpe and @bratseth PR

### CODE REVIEW

**Review by bratseth — APPROVED** [review]
Great!

File: container-search/src/main/java/com/yahoo/search/dispatch/rpc/RpcPing.java [hunk 1]
```diff
@@ -33,32 +31,41 @@ public RpcPing(Node node, ClusterMonitor<Node> clusterMonitor, RpcResourcePool r
...
-            } else if (responseOrError.error().isPresent()) {
-                return new Pong(ErrorMessage.createBackendCommunicationError(responseOrError.error().get()));
+            if (node.sendPing()) {
+                sendPing(handler);
             }
+        } catch (RuntimeException e) {
+            handler.handle(new Pong(
+                    ErrorMessage.createBackendCommunicationError("Exception when pinging " + node + ": " + Exceptions.toMessageString(e))));
+            node.receivePing();
+        } catch (Throwable throwable) {
```
Comment (havardpe) [inline container-search/src/main/java/com/yahoo/search/dispatch/rpc/RpcPing.java]: better to use finally than to hide Errors

File: container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/Node.java [hunk 2]
```diff
@@ -21,13 +21,20 @@
...
+    private final AtomicBoolean pendingPing = new AtomicBoolean();
 
     public Node(int key, String hostname, int group) {
         this.key = key;
         this.hostname = hostname;
         this.group = group;
     }
 
+    /** Only send ping if this method return true. If not the is a ping outstanding. */
+    public boolean sendPing() { return ! pendingPing.getAndSet(true); }
```
Comment (havardpe) [inline container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/Node.java]: somewhat misleading name

File: container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/Node.java [hunk 3]
```diff
@@ -21,13 +21,20 @@
...
+
+    /** Need to be called when a pong is called to allow next ping to go through. */
+    public void receivePing() { pendingPing.set(false); }
```
Comment (havardpe) [inline container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/Node.java]: somewhat misleading name

**Review by havardpe — COMMENTED** [review]
seems simpler and better, a good combination
please consider my comments

File: container-search/src/main/java/com/yahoo/search/dispatch/rpc/RpcPing.java (hunk 1 above)
Comment (baldersheim) [reply container-search/src/main/java/com/yahoo/search/dispatch/rpc/RpcPing.java]: Take another look, Rewritten to use sequence numbers instead. Simpler and safer.

File: container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/Node.java (hunk 2 above)
Comment (baldersheim) [reply container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/Node.java]: Take another look, Rewritten to use sequence numbers instead. Simpler and safer.

File: container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/Node.java (hunk 3 above)
Comment (baldersheim) [reply container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/Node.java]: Take another look, Rewritten to use sequence numbers instead. Simpler and safer.

File: container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/Node.java [hunk 4]
```diff
@@ -21,19 +21,26 @@
     private final AtomicBoolean statusIsKnown = new AtomicBoolean(false);
     private final AtomicBoolean working = new AtomicBoolean(true);
     private final AtomicLong activeDocuments = new AtomicLong(0);
-    private final AtomicBoolean pendingPing = new AtomicBoolean();
+    private final AtomicLong pingSequence = new AtomicLong(0);
+    private final AtomicLong lastPing = new AtomicLong(0);
```
Comment (havardpe) [inline container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/Node.java]: pong?

File: container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/Node.java [hunk 5]
```diff
@@ -21,19 +21,26 @@
...
-
-    /** Need to be called when a pong is called to allow next ping to go through. */
-    public void receivePing() { pendingPing.set(false); }
+    /** Give a monotonically increasing sequence number.*/
+    public long createPingSequenceId() { return pingSequence.incrementAndGet(); }
+    /** Checks if this pong is received in line and accepted, or out of band and should be ignored..*/
+    public boolean isLastReceivedPong(long pingId ) {
+        long last = lastPing.get();
+        while ((pingId > last) && ! lastPing.compareAndSet(last, pingId)) {
+            last = pingSequence.get();
```
Comment (havardpe) [inline container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/Node.java]: mixing sent and received ping sequence number here seems like a bad idea

File: container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/Node.java (hunk 4 above)
Comment (baldersheim) [reply container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/Node.java]: Fixed

File: container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/Node.java (hunk 5 above)
Comment (baldersheim) [reply container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/Node.java]: Fixed.

**Pull request merged**